File: LeetCode/python/LinkedListCycle.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    # @param head, a ListNode
    # @return a boolean
    def hasCycle(self, head):
        if(head == None or head.next == None):
            return False
			
        slowNode = head #每次指向下一个节点
        fastNode = head.next #每次指向下一个节点的下一个节点
        while(slowNode != None and fastNode != None):
            if(slowNode == fastNode):
                return True
			
            slowNode = slowNode.next
            if(fastNode.next != None):
                fastNode = fastNode.next.next
            else:
                return False
		
        return False
		

#测试程序
l1 = ListNode(1)
l2 = ListNode(2)
l3 = ListNode(3)
l4 = ListNode(4)
l5 = ListNode(5)

l1.next = l2
l2.next = l3
l3.next = l4
l4.next = l5
l5.next = l1 #Cycle

# head = l1
# while(head != None):
#     print(head.val)
#     head = head.next

s = Solution()
print(s.hasCycle(l1))			 
		

File: LeetCode/c++/LinkedListCycle.cpp
#include <iostream>

using namespace std;

//Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

bool hasCycle(ListNode *head) 
{
	if (head == NULL || head->next == NULL)
		return false;

	ListNode *slowNode = head;
	ListNode *fastNode = head->next;

	while (slowNode != NULL && fastNode != NULL)
	{
		if (slowNode == fastNode)
			return true;

		slowNode = slowNode->next;
		if (fastNode->next != NULL)
			fastNode = fastNode->next->next;
		else
			break;
	}

	return false;
}

int main()
{
	ListNode *n1 = new ListNode(1);
	ListNode *n2 = new ListNode(2);
	ListNode *n3 = new ListNode(3);
	ListNode *n4 = new ListNode(4);
	ListNode *n5 = new ListNode(5);

	n1->next = n2;
	n2->next = n3;
	n3->next = n4;
	n4->next = n5;
	n5->next = n5;

	if (hasCycle(n1))
		cout << "Yes";
	else
		cout << "No";

	return 0;
}